llvm/lib/Support/Mustache.cpp

```diff
@@ -0,0 +1,489 @@
+// is represented only by {{& variable}}
+SmallVector<Token, 0> tokenize(StringRef Template) {
+  SmallVector<Token, 0> Tokens;
+  StringRef Open("{{");
+  StringRef Close("}}");
+  std::size_t Start = 0;
+  std::size_t DelimiterStart = Template.find(Open);
+  if (DelimiterStart == StringRef::npos) {
+    Tokens.emplace_back(Template);
+    return Tokens;
+  }
+  while (DelimiterStart != StringRef::npos) {
+    if (DelimiterStart != Start) {
+      Tokens.emplace_back(Template.substr(Start, DelimiterStart - Start));
+    }
+
+    size_t DelimiterEnd = Template.find(Close, DelimiterStart);
+    if (DelimiterEnd == StringRef::npos) {
+      break;
+    }
+
+    // Extract the Interpolated variable without {{ and }}
+    size_t InterpolatedStart = DelimiterStart + Open.size();
+    size_t InterpolatedEnd = DelimiterEnd - DelimiterStart - Close.size();
+    SmallString<128> Interpolated =
+        Template.substr(InterpolatedStart, InterpolatedEnd);
+    SmallString<128> RawBody;
+    RawBody += Open;
+    RawBody += Interpolated;
+    RawBody += Close;
+
+    Tokens.emplace_back(RawBody, Interpolated, Interpolated[0]);
+    Start = DelimiterEnd + Close.size();
+    DelimiterStart = Template.find(Open, Start);
+  }
+
+  if (Start < Template.size())
+    Tokens.emplace_back(Template.substr(Start));
+
+  // fix up white spaces for
+  // open sections/inverted sections/close section/comment
+  // This loop attempts to find standalone tokens and tries to trim out
+  // the whitespace around them
+  // for example:
+  // if you have the template string
+  //  "Line 1\n {{#section}} \n Line 2 \n {{/section}} \n Line 3"
+  // The output would be
+  //  "Line 1\n Line 2\n Line 3"
+  size_t LastIdx = Tokens.size() - 1;
+  for (size_t Idx = 0, End = Tokens.size(); Idx < End; ++Idx) {
+    Token &CurrentToken = Tokens[Idx];
+    Token::Type CurrentType = CurrentToken.getType();
+    // Check if token type requires cleanup
+    bool RequiresCleanUp = (CurrentType == Token::Type::SectionOpen ||
+                            CurrentType == Token::Type::InvertSectionOpen ||
+                            CurrentType == Token::Type::SectionClose ||
+                            CurrentType == Token::Type::Comment ||
+                            CurrentType == Token::Type::Partial);
+
+    if (!RequiresCleanUp)
+      continue;
+
+    bool NoTextBehind = noTextBehind(Idx, Tokens);
+    bool NoTextAhead = noTextAhead(Idx, Tokens);
+
+    // Adjust next token body if no text ahead
+    if ((NoTextBehind && NoTextAhead) || (NoTextAhead && Idx == 0)) {
+      Token &NextToken = Tokens[Idx + 1];
+      StringRef NextTokenBody = NextToken.getTokenBody();
+      if (NextTokenBody.starts_with("\r\n")) {
+        NextToken.setTokenBody(NextTokenBody.substr(2));
+      } else if (NextTokenBody.starts_with("\n")) {
+        NextToken.setTokenBody(NextTokenBody.substr(1));
+      }
+    }
+    // Adjust previous token body if no text behind
```

Comment:
Please add a more complete description here. While its obvious you're adjusting something, it isn't clear why or how.

Maybe these should be helper functions? They're not used elsewhere, but sometimes it helps to separate the logic.